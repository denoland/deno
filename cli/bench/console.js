// Copyright 2018-2024 the Deno authors. All rights reserved. MIT license.

// deno-lint-ignore-file no-console

const count = 100000;
for (let i = 0; i < count; i++) console.log("Hello World");
