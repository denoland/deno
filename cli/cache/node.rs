// Copyright 2018-2025 the Deno authors. MIT license.

use deno_core::error::AnyError;
use deno_core::serde_json;
use deno_runtime::deno_webstorage::rusqlite::params;

use super::cache_db::CacheDB;
use super::cache_db::CacheDBConfiguration;
use super::cache_db::CacheFailure;
use super::CacheDBHash;
use crate::node::CliCjsAnalysis;

pub static NODE_ANALYSIS_CACHE_DB: CacheDBConfiguration =
  CacheDBConfiguration {
    table_initializer: concat!(
      "CREATE TABLE IF NOT EXISTS cjsanalysiscache (",
      "specifier TEXT PRIMARY KEY,",
      "source_hash INTEGER NOT NULL,",
      "data TEXT NOT NULL",
      ");"
    ),
    on_version_change: "DELETE FROM cjsanalysiscache;",
    preheat_queries: &[],
    on_failure: CacheFailure::InMemory,
  };

#[derive(Clone)]
pub struct NodeAnalysisCache {
  inner: NodeAnalysisCacheInner,
}

impl NodeAnalysisCache {
  pub fn new(db: CacheDB) -> Self {
    Self {
      inner: NodeAnalysisCacheInner::new(db),
    }
  }

  fn ensure_ok<T: Default>(res: Result<T, AnyError>) -> T {
    match res {
      Ok(x) => x,
      Err(err) => {
        // TODO(mmastrac): This behavior was inherited from before the refactoring but it probably makes sense to move it into the cache
        // at some point.
        // should never error here, but if it ever does don't fail
        if cfg!(debug_assertions) {
          panic!("Error using esm analysis: {err:#}");
        } else {
          log::debug!("Error using esm analysis: {:#}", err);
        }
        T::default()
      }
    }
  }

  pub fn get_cjs_analysis(
    &self,
    specifier: &str,
    expected_source_hash: CacheDBHash,
  ) -> Option<CliCjsAnalysis> {
    Self::ensure_ok(
      self.inner.get_cjs_analysis(specifier, expected_source_hash),
    )
  }

  pub fn set_cjs_analysis(
    &self,
    specifier: &str,
    source_hash: CacheDBHash,
    cjs_analysis: &CliCjsAnalysis,
  ) {
    Self::ensure_ok(self.inner.set_cjs_analysis(
      specifier,
      source_hash,
      cjs_analysis,
    ));
  }
}

#[derive(Clone)]
struct NodeAnalysisCacheInner {
  conn: CacheDB,
}

impl NodeAnalysisCacheInner {
  pub fn new(conn: CacheDB) -> Self {
    Self { conn }
  }

  pub fn get_cjs_analysis(
    &self,
    specifier: &str,
    expected_source_hash: CacheDBHash,
  ) -> Result<Option<CliCjsAnalysis>, AnyError> {
    let query = "
      SELECT
        data
      FROM
        cjsanalysiscache
      WHERE
        specifier=?1
        AND source_hash=?2
      LIMIT 1";
    let res = self.conn.query_row(
      query,
      params![specifier, expected_source_hash],
      |row| {
        let analysis_info: String = row.get(0)?;
        Ok(serde_json::from_str(&analysis_info)?)
      },
    )?;
    Ok(res)
  }

  pub fn set_cjs_analysis(
    &self,
    specifier: &str,
    source_hash: CacheDBHash,
    cjs_analysis: &CliCjsAnalysis,
  ) -> Result<(), AnyError> {
    let sql = "
      INSERT OR REPLACE INTO
        cjsanalysiscache (specifier, source_hash, data)
      VALUES
        (?1, ?2, ?3)";
    self.conn.execute(
      sql,
      params![
        specifier,
        source_hash,
        &serde_json::to_string(&cjs_analysis)?,
      ],
    )?;
    Ok(())
  }
}

#[cfg(test)]
mod test {
  use super::*;

  #[test]
  pub fn node_analysis_cache_general_use() {
    let conn = CacheDB::in_memory(&NODE_ANALYSIS_CACHE_DB, "1.0.0");
    let cache = NodeAnalysisCacheInner::new(conn);

    assert!(cache
      .get_cjs_analysis("file.js", CacheDBHash::new(2))
      .unwrap()
      .is_none());
    let cjs_analysis = CliCjsAnalysis::Cjs {
      exports: vec!["export1".to_string()],
      reexports: vec!["re-export1".to_string()],
    };
    cache
      .set_cjs_analysis("file.js", CacheDBHash::new(2), &cjs_analysis)
      .unwrap();
    assert!(cache
      .get_cjs_analysis("file.js", CacheDBHash::new(3))
      .unwrap()
      .is_none()); // different hash
    let actual_cjs_analysis = cache
      .get_cjs_analysis("file.js", CacheDBHash::new(2))
      .unwrap()
      .unwrap();
    assert_eq!(actual_cjs_analysis, cjs_analysis);

    // adding when already exists should not cause issue
    cache
      .set_cjs_analysis("file.js", CacheDBHash::new(2), &cjs_analysis)
      .unwrap();

    // recreating with same cli version should still have it
    let conn = cache.conn.recreate_with_version("1.0.0");
    let cache = NodeAnalysisCacheInner::new(conn);
    let actual_analysis = cache
      .get_cjs_analysis("file.js", CacheDBHash::new(2))
      .unwrap()
      .unwrap();
    assert_eq!(actual_analysis, cjs_analysis);

    // now changing the cli version should clear it
    let conn = cache.conn.recreate_with_version("2.0.0");
    let cache = NodeAnalysisCacheInner::new(conn);
    assert!(cache
      .get_cjs_analysis("file.js", CacheDBHash::new(2))
      .unwrap()
      .is_none());
  }
}
