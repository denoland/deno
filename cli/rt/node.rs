// Copyright 2018-2025 the Deno authors. MIT license.

use std::borrow::Cow;
use std::sync::Arc;

use deno_core::url::Url;
use deno_error::JsErrorBox;
use deno_lib::loader::NpmModuleLoader;
use deno_lib::standalone::binary::CjsExportAnalysisEntry;
use deno_media_type::MediaType;
use deno_resolver::npm::DenoInNpmPackageChecker;
use deno_resolver::npm::NpmReqResolver;
use deno_runtime::deno_fs::FileSystem;
use node_resolver::analyze::CjsAnalysis;
use node_resolver::analyze::CjsAnalysisExports;
use node_resolver::analyze::NodeCodeTranslator;
use node_resolver::DenoIsBuiltInNodeModuleChecker;

use crate::binary::StandaloneModules;
use crate::file_system::DenoRtSys;

pub type DenoRtCjsTracker =
  deno_resolver::cjs::CjsTracker<DenoInNpmPackageChecker, DenoRtSys>;
pub type DenoRtNpmResolver = deno_resolver::npm::NpmResolver<DenoRtSys>;
pub type DenoRtNpmModuleLoader = NpmModuleLoader<
  CjsCodeAnalyzer,
  DenoInNpmPackageChecker,
  DenoIsBuiltInNodeModuleChecker,
  DenoRtNpmResolver,
  DenoRtSys,
>;
pub type DenoRtNodeCodeTranslator = NodeCodeTranslator<
  CjsCodeAnalyzer,
  DenoInNpmPackageChecker,
  DenoIsBuiltInNodeModuleChecker,
  DenoRtNpmResolver,
  DenoRtSys,
>;
pub type DenoRtNodeResolver = deno_runtime::deno_node::NodeResolver<
  DenoInNpmPackageChecker,
  DenoRtNpmResolver,
  DenoRtSys,
>;
pub type DenoRtNpmReqResolver = NpmReqResolver<
  DenoInNpmPackageChecker,
  DenoIsBuiltInNodeModuleChecker,
  DenoRtNpmResolver,
  DenoRtSys,
>;

pub struct CjsCodeAnalyzer {
  cjs_tracker: Arc<DenoRtCjsTracker>,
  modules: Arc<StandaloneModules>,
  sys: DenoRtSys,
}

impl CjsCodeAnalyzer {
  pub fn new(
    cjs_tracker: Arc<DenoRtCjsTracker>,
    modules: Arc<StandaloneModules>,
    sys: DenoRtSys,
  ) -> Self {
    Self {
      cjs_tracker,
      modules,
      sys,
    }
  }

  fn inner_cjs_analysis<'a>(
    &self,
    specifier: &Url,
    source: Cow<'a, str>,
  ) -> Result<CjsAnalysis<'a>, JsErrorBox> {
    let media_type = MediaType::from_specifier(specifier);
    if media_type == MediaType::Json {
      return Ok(CjsAnalysis::Cjs(CjsAnalysisExports {
        exports: vec![],
        reexports: vec![],
      }));
    }

    let cjs_tracker = self.cjs_tracker.clone();
    let is_maybe_cjs = cjs_tracker
      .is_maybe_cjs(specifier, media_type)
      .map_err(JsErrorBox::from_err)?;
    let analysis = if is_maybe_cjs {
      let data = self
        .modules
        .read(specifier)?
        .and_then(|d| d.cjs_export_analysis);
      match data {
        Some(data) => {
          let data: CjsExportAnalysisEntry = bincode::deserialize(&data)
            .map_err(|err| JsErrorBox::generic(err.to_string()))?;
          match data {
            CjsExportAnalysisEntry::Esm => {
              cjs_tracker.set_is_known_script(specifier, false);
              CjsAnalysis::Esm(source)
            }
            CjsExportAnalysisEntry::Cjs(analysis) => {
              cjs_tracker.set_is_known_script(specifier, true);
              CjsAnalysis::Cjs(analysis)
            }
          }
        }
        None => {
          if log::log_enabled!(log::Level::Debug) {
            if self.sys.is_specifier_in_vfs(specifier) {
              log::debug!(
                "No CJS export analysis was stored for '{}'. Assuming ESM. This might indicate a bug in Deno.",
                specifier
              );
            } else {
              log::debug!(
                "Analyzing potentially CommonJS files is not supported at runtime in a compiled executable ({}). Assuming ESM.",
                specifier
              );
            }
          }
          // assume ESM as we don't have access to swc here
          CjsAnalysis::Esm(source)
        }
      }
    } else {
      CjsAnalysis::Esm(source)
    };

    Ok(analysis)
  }
}

#[async_trait::async_trait(?Send)]
impl node_resolver::analyze::CjsCodeAnalyzer for CjsCodeAnalyzer {
  async fn analyze_cjs<'a>(
    &self,
    specifier: &Url,
    source: Option<Cow<'a, str>>,
  ) -> Result<CjsAnalysis<'a>, JsErrorBox> {
    let source = match source {
      Some(source) => source,
      None => {
        if let Ok(path) = deno_path_util::url_to_file_path(specifier) {
          // todo(dsherret): should this use the sync method instead?
          if let Ok(source_from_file) =
            self.sys.read_text_file_lossy_async(path, None).await
          {
            source_from_file
          } else {
            return Ok(CjsAnalysis::Cjs(CjsAnalysisExports {
              exports: vec![],
              reexports: vec![],
            }));
          }
        } else {
          return Ok(CjsAnalysis::Cjs(CjsAnalysisExports {
            exports: vec![],
            reexports: vec![],
          }));
        }
      }
    };
    self.inner_cjs_analysis(specifier, source)
  }
}
