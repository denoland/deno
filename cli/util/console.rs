// Copyright 2018-2025 the Deno authors. MIT license.

use deno_runtime::ops::tty::ConsoleSize;

/// Gets the console size.
pub fn console_size() -> Option<ConsoleSize> {
  let stderr = &deno_runtime::deno_io::STDERR_HANDLE;
  deno_runtime::ops::tty::console_size(stderr).ok()
}
