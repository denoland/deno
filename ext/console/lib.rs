// Copyright 2018-2025 the Deno authors. MIT license.
use std::path::PathBuf;

use deno_core::op2;
use deno_core::v8;

deno_core::extension!(
  deno_console,
  ops = [op_preview_entries],
  esm = ["01_console.js"],
);

pub fn get_declaration() -> PathBuf {
  PathBuf::from(env!("CARGO_MANIFEST_DIR")).join("lib.deno_console.d.ts")
}

#[op2]
pub fn op_preview_entries<'s>(
  scope: &mut v8::HandleScope<'s>,
  object: &v8::Object,
  slow_path: bool,
) -> v8::Local<'s, v8::Value> {
  let (entries, is_key_value) = object.preview_entries(scope);
  match entries {
    None => v8::undefined(scope).into(),
    Some(entries) => {
      if !slow_path {
        return entries.into();
      }

      let ret: [v8::Local<v8::Value>; 2] =
        [entries.into(), v8::Boolean::new(scope, is_key_value).into()];
      v8::Array::new_with_elements(scope, &ret).into()
    }
  }
}
