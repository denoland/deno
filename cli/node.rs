// Copyright 2018-2024 the Deno authors. All rights reserved. MIT license.

use std::borrow::Cow;
use std::sync::Arc;

use deno_ast::MediaType;
use deno_ast::ModuleSpecifier;
use deno_core::error::AnyError;
use deno_graph::ParsedSourceStore;
use deno_runtime::deno_fs;
use deno_runtime::deno_node::DenoFsNodeResolverEnv;
use node_resolver::analyze::CjsAnalysis as ExtNodeCjsAnalysis;
use node_resolver::analyze::CjsAnalysisExports;
use node_resolver::analyze::CjsCodeAnalyzer;
use node_resolver::analyze::NodeCodeTranslator;
use serde::Deserialize;
use serde::Serialize;

use crate::cache::CacheDBHash;
use crate::cache::NodeAnalysisCache;
use crate::cache::ParsedSourceCache;
use crate::resolver::CjsTracker;

pub type CliNodeCodeTranslator =
  NodeCodeTranslator<CliCjsCodeAnalyzer, DenoFsNodeResolverEnv>;

/// Resolves a specifier that is pointing into a node_modules folder.
///
/// Note: This should be called whenever getting the specifier from
/// a Module::External(module) reference because that module might
/// not be fully resolved at the time deno_graph is analyzing it
/// because the node_modules folder might not exist at that time.
pub fn resolve_specifier_into_node_modules(
  specifier: &ModuleSpecifier,
  fs: &dyn deno_fs::FileSystem,
) -> ModuleSpecifier {
  node_resolver::resolve_specifier_into_node_modules(specifier, &|path| {
    fs.realpath_sync(path).map_err(|err| err.into_io_error())
  })
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub enum CliCjsAnalysis {
  /// The module was found to be an ES module.
  Esm,
  /// The module was CJS.
  Cjs {
    exports: Vec<String>,
    reexports: Vec<String>,
  },
}

pub struct CliCjsCodeAnalyzer {
  cache: NodeAnalysisCache,
  cjs_tracker: Arc<CjsTracker>,
  fs: deno_fs::FileSystemRc,
  parsed_source_cache: Option<Arc<ParsedSourceCache>>,
}

impl CliCjsCodeAnalyzer {
  pub fn new(
    cache: NodeAnalysisCache,
    cjs_tracker: Arc<CjsTracker>,
    fs: deno_fs::FileSystemRc,
    parsed_source_cache: Option<Arc<ParsedSourceCache>>,
  ) -> Self {
    Self {
      cache,
      cjs_tracker,
      fs,
      parsed_source_cache,
    }
  }

  async fn inner_cjs_analysis(
    &self,
    specifier: &ModuleSpecifier,
    source: &str,
  ) -> Result<CliCjsAnalysis, AnyError> {
    let source_hash = CacheDBHash::from_source(source);
    if let Some(analysis) =
      self.cache.get_cjs_analysis(specifier.as_str(), source_hash)
    {
      return Ok(analysis);
    }

    let media_type = MediaType::from_specifier(specifier);
    if media_type == MediaType::Json {
      return Ok(CliCjsAnalysis::Cjs {
        exports: vec![],
        reexports: vec![],
      });
    }

    let cjs_tracker = self.cjs_tracker.clone();
    let is_maybe_cjs = cjs_tracker.is_maybe_cjs(specifier, media_type)?;
    let analysis = if is_maybe_cjs {
      let maybe_parsed_source = self
        .parsed_source_cache
        .as_ref()
        .and_then(|c| c.remove_parsed_source(specifier));

      deno_core::unsync::spawn_blocking({
        let specifier = specifier.clone();
        let source: Arc<str> = source.into();
        move || -> Result<_, AnyError> {
          let parsed_source =
            maybe_parsed_source.map(Ok).unwrap_or_else(|| {
              deno_ast::parse_program(deno_ast::ParseParams {
                specifier,
                text: source,
                media_type,
                capture_tokens: true,
                scope_analysis: false,
                maybe_syntax: None,
              })
            })?;
          let is_script = parsed_source.compute_is_script();
          let is_cjs = cjs_tracker.is_cjs_with_known_is_script(
            parsed_source.specifier(),
            media_type,
            is_script,
          )?;
          if is_cjs {
            let analysis = parsed_source.analyze_cjs();
            Ok(CliCjsAnalysis::Cjs {
              exports: analysis.exports,
              reexports: analysis.reexports,
            })
          } else {
            Ok(CliCjsAnalysis::Esm)
          }
        }
      })
      .await
      .unwrap()?
    } else {
      CliCjsAnalysis::Esm
    };

    self
      .cache
      .set_cjs_analysis(specifier.as_str(), source_hash, &analysis);

    Ok(analysis)
  }
}

#[async_trait::async_trait(?Send)]
impl CjsCodeAnalyzer for CliCjsCodeAnalyzer {
  async fn analyze_cjs<'a>(
    &self,
    specifier: &ModuleSpecifier,
    source: Option<Cow<'a, str>>,
  ) -> Result<ExtNodeCjsAnalysis<'a>, AnyError> {
    let source = match source {
      Some(source) => source,
      None => {
        if let Ok(path) = specifier.to_file_path() {
          if let Ok(source_from_file) =
            self.fs.read_text_file_lossy_async(path, None).await
          {
            Cow::Owned(source_from_file)
          } else {
            return Ok(ExtNodeCjsAnalysis::Cjs(CjsAnalysisExports {
              exports: vec![],
              reexports: vec![],
            }));
          }
        } else {
          return Ok(ExtNodeCjsAnalysis::Cjs(CjsAnalysisExports {
            exports: vec![],
            reexports: vec![],
          }));
        }
      }
    };
    let analysis = self.inner_cjs_analysis(specifier, &source).await?;
    match analysis {
      CliCjsAnalysis::Esm => Ok(ExtNodeCjsAnalysis::Esm(source)),
      CliCjsAnalysis::Cjs { exports, reexports } => {
        Ok(ExtNodeCjsAnalysis::Cjs(CjsAnalysisExports {
          exports,
          reexports,
        }))
      }
    }
  }
}
