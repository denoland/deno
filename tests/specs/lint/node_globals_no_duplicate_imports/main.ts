import {} from "node:console";

const _foo = setImmediate;

import {} from "node:assert";

const _bar = setImmediate;
