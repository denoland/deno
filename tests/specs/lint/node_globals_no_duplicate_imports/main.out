error: Uncaught (in promise) ReferenceError: setImmediate is not defined
const _foo = setImmediate;
             ^
    at [WILDCARD]main.ts:3:14

    info: setImmediate is not available in the global scope in Deno.
    hint: Import it explicitly with import { setImmediate } from "node:timers";,
          or run again with --unstable-node-globals flag to add this global.
